services/merak-agent/handler/create.go

```diff
@@ -0,0 +1,399 @@
+	The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+	WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+package handler
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	pb "github.com/futurewei-cloud/merak/api/proto/v1/agent"
+	common_pb "github.com/futurewei-cloud/merak/api/proto/v1/common"
+	constants "github.com/futurewei-cloud/merak/services/common"
+	"github.com/tidwall/gjson"
+)
+
+type port struct {
+	Port minimalPort `json:"port"`
+}
+
+type minimalPort struct {
+	AdminState bool                `json:"admin_state_up"`
+	DeviceID   string              `json:"device_id"`
+	NetworkID  string              `json:"network_id"`
+	SG         []string            `json:"security_groups"`
+	FixIPs     []map[string]string `json:"fixed_ips"`
+	TenantID   string              `json:"tenant_id"`
+}
+
+type updatePortMain struct {
+	Port updatePort `json:"port"`
+}
+
+type updatePort struct {
+	ProjectID     string `json:"project_id"`
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	Description   string `json:"description"`
+	NetworkID     string `json:"network_id"`
+	TenantID      string `json:"tenant_id"`
+	AdminState    bool   `json:"admin_state_up"`
+	VethName      string `json:"veth_name"`
+	DeviceID      string `json:"device_id"`
+	DeviceOwner   string `json:"device_owner"`
+	FastPath      bool   `json:"fast_path"`
+	BindingHostID string `json:"binding:host_id"`
+}
+
+func caseCreate(ctx context.Context, in *pb.InternalPortConfig) (*pb.AgentReturnInfo, error) {
+
+	log.Println("Create Minimal Port")
+	vmInfo := pb.ReturnPortInfo{
+		Ip:       "",
+		Deviceid: "",
+		Remoteid: "",
+		Status:   common_pb.Status_ERROR,
+	}
+	minimalPortBody := port{
+		Port: minimalPort{
+			AdminState: true,
+			DeviceID:   in.Name,
+			NetworkID:  in.Vpcid,
+			SG:         []string{in.Sg},
+			FixIPs:     []map[string]string{{"subnet_id": in.Subnetid}},
+			TenantID:   in.Tenantid,
+		},
+	}
+
+	body, err := json.Marshal(minimalPortBody)
+	if err != nil {
+		return &pb.AgentReturnInfo{
+			ReturnMessage: "Failed to marshal json!",
+			ReturnCode:    common_pb.ReturnCode_FAILED,
+			Port:          &vmInfo,
+		}, err
+	}
+	log.Println("Sending body to Alcor: \n", string(body[:]))
+	resp, err := http.Post("http://"+constants.ALCOR_ADDRESS+":"+strconv.Itoa(constants.ALCOR_PORT_MANAGER_PORT)+"/project/"+in.Projectid+"/ports", "application/json", bytes.NewBuffer(body))
```

Comment:
@phudtran Are we still hard code for `ALCOR_ADDRESS`?
Scenario Manager already has Service-config for all urls including port-create, port-update, port-delete.
We should get the urls from the protobuf in Service-config.

Reply:
I was thinking of having the test target address set in a config map for the cluster.
That way our components can read from it on startup, rather than have the scenario manager send it in every message.
What do you think?

Reply:
@phudtran I agree. Please create an issue for it. I think we need to let Merak-topology to get alcor's url information from services config in protobuf and create a config map for merak-agent service. Please discuss with @zzengcs about this issue and let her know how to create such config map for merak-agent.